src/p11_child/p11_child_openssl.c

```diff
@@ -0,0 +1,1133 @@
+    }
+    pkcs11_session = true;
+
+    /* login: do we need to check for Login Required? */
+    if (mode == OP_AUTH) {
+        DEBUG(SSSDBG_TRACE_ALL, "Login required.\n");
+        if (pin != NULL) {
+            rv = module->C_Login(session, CKU_USER, discard_const(pin),
+                                 strlen(pin));
+            if (rv != CKR_OK) {
+                DEBUG(SSSDBG_OP_FAILURE, "C_Login failed.\n");
+                ret = EIO;
+                goto done;
+            }
+            pkcs11_login = true;
+        } else {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Login required but no PIN available, continue.\n");
+        }
+    } else {
+        DEBUG(SSSDBG_TRACE_ALL, "Login NOT required.\n");
+    }
+
+    ret = read_certs(mem_ctx, module, session, p11_ctx, &cert_list);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "read_certs failed.\n");
+        goto done;
+    }
+
+    /* TODO: check module_name_in, token_name_in, key_id_in */
+
+    if (cert_list == NULL) {
+        DEBUG(SSSDBG_TRACE_ALL, "No certificate found.\n");
+        *_multi = NULL;
+        ret = EOK;
+        goto done;
+    }
+
+    if (mode == OP_AUTH) {
+        if (cert_list->next != NULL || cert_list->prev != NULL) {
+            DEBUG(SSSDBG_FATAL_FAILURE,
+                  "More than one certificate found for authentication, "
+                  "aborting!\n");
+            ret = EINVAL;
+            goto done;
+        }
+
+        ret = sign_data(module, session, cert_list);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "sign_data failed.\n");
+            ret = EACCES;
+            goto done;
+        }
+
+        DEBUG(SSSDBG_TRACE_ALL,
+              "Certificate verified and validated.\n");
+    }
+
+    multi = talloc_strdup(mem_ctx, "");
+    if (multi == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Failed to create output string.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    DLIST_FOR_EACH(item, cert_list) {
+        DEBUG(SSSDBG_TRACE_ALL, "Found certificate has key id [%s].\n",
+              item->id);
+
+        multi = talloc_asprintf_append(multi, "%s\n%s\n%s\n%s\n%s\n",
+                                       token_name, module_file_name, item->id,
+                                       item->label, item->cert_b64);
+    }
+
+    *_multi = multi;
+
+    ret = EOK;
+done:
+    if (module != NULL) {
+        if (pkcs11_login) {
+            rv = module->C_Logout(session);
+            if (rv != CKR_OK) {
+                DEBUG(SSSDBG_OP_FAILURE, "C_Logout failed.\n");
+            }
+        }
+        if (pkcs11_session) {
+            rv = module->C_CloseSession(session);
+            if (rv != CKR_OK) {
+                DEBUG(SSSDBG_OP_FAILURE, "C_CloseSession failed.\n");
+            }
+        }
+    }
+    free(slot_name);
+    free(token_name);
+    free(module_file_name);
+    p11_kit_modules_finalize_and_release(modules);
+
+    return ret;
+}
+
+#if 0
```

Comment:
Another if-ed block

Reply:
I'm sorry for the if-block, I must have checked the _nss file twice but forgot the _openssl version.